File: lib/core/pump_guard.dart
/// 펌프 프리셋 clamp + 부피→스텝 파생 — SoT §6 (byte-parity 안전 급소).
///
/// **바이트 동일 축 = 서버 `settingsClamp.ts`(TS·PumpGuard 부분) ↔ 이 파일(Dart).**
/// P0 HW 안전 — 식향 Code 11(플런저 오버로드·과다흡입) 재발 방지. 근본원인은
/// "하드코딩 24000 vs 파생 9600"의 2.5배 불일치였다(SoT 서두).
///
/// ⚠️ SoT §6 서두: 현재 web 에 이 축의 서버쪽 절반이 빠져 있다. 이 Dart 포트는 **SoT §6 수치표·
///   알고리즘·라운딩·단조성 보정 순서**를 정본으로 삼는다. 서버 복원 시 이 상수·순서에 맞춘다.
///   (부록A P-7: clampPumpPreset custom 단조성 2줄 순서 / P-8: steps=round half-up·fragrance ×1000)
library;

/// PumpPreset 7필드 — SoT §6-1 (고정 필드명·타입·순서).
class PumpPreset {
  const PumpPreset({
    required this.pumpPresetId,
    required this.pumpFullStroke,
    required this.pumpMaxStartSpeedHz,
    required this.pumpMaxTopSpeedHz,
    required this.pumpMaxCutoffSpeedHz,
    required this.pumpMaxSlope,
    required this.pumpSyringeTypeCode,
  });

  /// ∈ {sy01b, cavro_xlp6000, cavro_xcalibur, custom}.
  final String pumpPresetId;

  /// 풀스트로크.
  final int pumpFullStroke;

  /// v 상한(start speed).
  final int pumpMaxStartSpeedHz;

  /// V 상한(top speed).
  final int pumpMaxTopSpeedHz;

  /// c 상한(cutoff speed).
  final int pumpMaxCutoffSpeedHz;

  /// L 상한(slope).
  final int pumpMaxSlope;

  /// 스톨 서브코드 U<code>.
  final int pumpSyringeTypeCode;
}

/// 빌트인 프리셋 정식 수치표 — SoT §6-2 (입력 무시·강제 · 바이트 동일 SoT).
///
/// XCalibur 의 v/V/c 는 미확정(§6-2 `*`, O-12) → **보수적 SY-01B 하한**(1000/6000/5400) 채택.
/// 속도 clamp 는 낮을수록 안전.
const Map<String, PumpPreset> pumpPresets = {
  'sy01b': PumpPreset(
    pumpPresetId: 'sy01b',
    pumpFullStroke: 12000,
    pumpMaxStartSpeedHz: 1000,
    pumpMaxTopSpeedHz: 6000,
    pumpMaxCutoffSpeedHz: 5400,
    pumpMaxSlope: 20,
    pumpSyringeTypeCode: 200,
  ),
  'cavro_xlp6000': PumpPreset(
    pumpPresetId: 'cavro_xlp6000',
    pumpFullStroke: 6000,
    pumpMaxStartSpeedHz: 8000,
    pumpMaxTopSpeedHz: 48000,
    pumpMaxCutoffSpeedHz: 21600,
    pumpMaxSlope: 20,
    pumpSyringeTypeCode: 200, // 서버 pumpGuard.ts U200 (v1.1.0 확정 — 구 이관본 0은 stale 버그, 2026-07-12 정정)
  ),
  'cavro_xcalibur': PumpPreset(
    pumpPresetId: 'cavro_xcalibur',
    pumpFullStroke: 3000,
    pumpMaxStartSpeedHz: 1000, // * SY-01B 하한(미확정·보수적)
    pumpMaxTopSpeedHz: 6000, // * SY-01B 하한
    pumpMaxCutoffSpeedHz: 5400, // * SY-01B 하한
    pumpMaxSlope: 20,
    pumpSyringeTypeCode: 200, // 서버 pumpGuard.ts U200 (v1.1.0 확정 — 구 이관본 0은 stale 버그, 2026-07-12 정정)
  ),
};

/// custom 절대상한 — SoT §6-3.
const int _customStrokeMin = 100;
const int _customStrokeMax = 96000;
const int _customSpeedMin = 1;
const int _customSpeedMax = 48000;
const int _customSlopeMin = 1;
const int _customSlopeMax = 40;
const int _customTypeMin = 0;
const int _customTypeMax = 999;

/// 유효 syringe 용량 이산값(mL) — v1.1.0 서버 pumpGuard.ts VALID_SYRINGE_ML 정본 **9종**.
///   (구 이관본 4종은 stale — 2026-07-12 서버 SoT에 맞춰 정정. Python 배포본과 동일.)
///
/// (double 은 primitive equality 가 없어 const Set 불가 — final 런타임 집합.)
final Set<double> validSyringeCapacitiesMl = <double>{
  0.025,
  0.05,
  0.1,
  0.25,
  0.5,
  1.0,
  1.25,
  2.5,
  5.0,
};

/// 정수 clamp(round 후 [min,max]) — TS `clampInt`. NaN/누락 → fallback.
///
/// round = half-up(양수 도메인 JS `Math.round` ↔ Dart `round()` 실질 동일·SoT §6-4/O-14).
int _clampInt(Object? v, int min, int max, int fallback) {
  final num? n = v is num ? v : num.tryParse('$v');
  if (n == null || !n.isFinite) return fallback;
  final int r = n.round();
  return r < min ? min : (r > max ? max : r);
}

/// clampPumpPreset(cfg) — SoT §6-3 (서버 ↔ pi 동일 알고리즘).
///
/// 1) builtin(sy01b|cavro_xlp6000|cavro_xcalibur) → 표의 정식 수치 그대로(입력 전부 무시).
/// 2) custom → 절대상한 clamp + 속도 단조성 강제(⚠️ 2줄 순서 고정·부록A P-7).
/// 3) unknown id → sy01b 폴백.
/// 4) 미설정/누락 → sy01b 프리셋(호출 측이 null 전달).
PumpPreset clampPumpPreset(Map<String, Object?>? cfg) {
  final Object? rawId = cfg?['pumpPresetId'];
  final String id = rawId is String ? rawId : 'sy01b';

  // 1) builtin — 정식 수치 강제(입력 수치 전부 무시).
  final PumpPreset? builtin = pumpPresets[id];
  if (builtin != null && id != 'custom') return builtin;

  // 3) unknown id(그리고 non-custom 이 아닌 알 수 없는 값) → sy01b 폴백.
  if (id != 'custom') return pumpPresets['sy01b']!;

  // 2) custom → 절대상한 clamp.
  final int stroke =
      _clampInt(cfg?['pumpFullStroke'], _customStrokeMin, _customStrokeMax, 12000);
  int v = _clampInt(cfg?['pumpMaxStartSpeedHz'], _customSpeedMin, _customSpeedMax, 1000);
  final int bigV = _clampInt(cfg?['pumpMaxTopSpeedHz'], _customSpeedMin, _customSpeedMax, 6000);
  int c = _clampInt(cfg?['pumpMaxCutoffSpeedHz'], _customSpeedMin, _customSpeedMax, 5400);
  final int slope = _clampInt(cfg?['pumpMaxSlope'], _customSlopeMin, _customSlopeMax, 20);
  final int typeCode = _clampInt(cfg?['pumpSyringeTypeCode'], _customTypeMin, _customTypeMax, 200);

  // 속도 단조성 강제 — ⚠️ 순서 고정(부록A P-7: 2줄 순서가 바이트 동일이어야 경계 입력 결과 일치).
  //   (SY-01B 제약 v ≤ c ≤ V)
  c = _min(_max(c, v), bigV);
  v = _min(v, c);

  return PumpPreset(
    pumpPresetId: 'custom',
    pumpFullStroke: stroke,
    pumpMaxStartSpeedHz: v,
    pumpMaxTopSpeedHz: bigV,
    pumpMaxCutoffSpeedHz: c,
    pumpMaxSlope: slope,
    pumpSyringeTypeCode: typeCode,
  );
}

int _min(int a, int b) => a < b ? a : b;
int _max(int a, int b) => a > b ? a : b;

/// syringeCapacityMl 이산값 검증 — SoT §6-1 / O-15.
///
/// 유효집합(1.25/0.5/2.5/5) 밖이면 **모드 기본값 폴백**(스냅 아님). flavor 1.25 / fragrance 0.5.
double resolveSyringeCapacityMl(Object? raw, {required bool isFlavor}) {
  final double fallback = isFlavor ? 1.25 : 0.5;
  if (raw is! num) return fallback;
  final double d = raw.toDouble();
  return validSyringeCapacitiesMl.contains(d) ? d : fallback;
}

/// 부피→스텝 파생(SyringeSpec) — SoT §6-4 (하드코딩 금지·파생이 SoT).
///
///   steps         = round( pumpFullStroke × volumeUl ÷ (syringeCapacityMl × 1000) )
///   stepsPerMl    = pumpFullStroke ÷ syringeCapacityMl
///   maxVolumeUl   = syringeCapacityMl × 1000   (per-pump 안전 게이트 상한)
///
/// 검산(§6-4): 12000 × 100 ÷ 1250 = 960 steps / 식향 1.25mL stepsPerMl=9600 /
///            향장향 0.5mL stepsPerMl=24000.
class SyringeSpec {
  SyringeSpec({required this.pumpFullStroke, required this.syringeCapacityMl});

  final int pumpFullStroke;
  final double syringeCapacityMl;

  /// per-pump 안전 게이트 상한(µL).
  double get maxVolumeUl => syringeCapacityMl * 1000;

  /// mL 당 스텝수.
  double get stepsPerMl => pumpFullStroke / syringeCapacityMl;

  /// 부피(µL) → 스텝수. round = half-up(양수 도메인·부록A P-8).
  int stepsForVolumeUl(double volumeUl) =>
      (pumpFullStroke * volumeUl / (syringeCapacityMl * 1000)).round();
}

/// fragrance 단위 정규화 — SoT §6-6. amountMl → volumeUl(µL). 미스매치 = Code 11.
/// (flavor volume 은 이미 µL 이므로 정규화 불필요.)
double fragranceMlToUl(double amountMl) => amountMl * 1000;

/// recipe 스텝 검증 게이트 — SoT §6-4 / §9-1.
///   0 < volumeUl ≤ maxVolumeUl · steps.length(=steps) ≥ 1. 위반 → CMD_VALIDATION_FAILED(drop).
bool isVolumeWithinGate(double volumeUl, SyringeSpec spec) =>
    volumeUl > 0 && volumeUl <= spec.maxVolumeUl;

/// EnginePort 에러코드 분류 — SoT §6-7.
enum EngineErrorClass { normal, transient, permanent }

/// 엔진 raw errorCode(정수) → 분류 — SoT §6-7.
///   0 = 정상 / 1·7·11·15·timeout = transient(R=3 재시도) / 2·3·9·10 = permanent(즉시중단 FAILED).
EngineErrorClass classifyEngineErrorCode(int code) {
  if (code == 0) return EngineErrorClass.normal;
  if (code == 1 || code == 7 || code == 11 || code == 15) return EngineErrorClass.transient;
  if (code == 2 || code == 3 || code == 9 || code == 10) return EngineErrorClass.permanent;
  // 미분류 코드는 보수적으로 permanent(안전측·즉시중단).
  return EngineErrorClass.permanent;
}

/// status.errorCode 7종 — SoT §6-7 / §9-2.
enum StatusErrorCode {
  cmdValidationFailed('CMD_VALIDATION_FAILED'),
  duplicateDropped('DUPLICATE_DROPPED'),
  engineTimeout('ENGINE_TIMEOUT'),
  engineErrorTransient('ENGINE_ERROR_TRANSIENT'),
  engineErrorPermanent('ENGINE_ERROR_PERMANENT'),
  partialDispense('PARTIAL_DISPENSE'),
  interrupted('INTERRUPTED');

  const StatusErrorCode(this.wire);
  final String wire;

  static StatusErrorCode? fromWire(Object? v) {
    if (v is! String) return null;
    for (final e in StatusErrorCode.values) {
      if (e.wire == v) return e;
    }
    return null;
  }
}

File: src/senlyt_pi/core/pump_guard.py
"""펌프 프리셋 clamp + 부피→스텝 파생 — SoT §6 (byte-parity 안전 급소).

**바이트 동일 축 = 서버 `pumpGuard.ts`/`settingsClamp.ts`(TS) ↔ 이 파일(Python).**
P0 HW 안전 — 식향 Code 11(플런저 오버로드·과다흡입) 재발 방지. 근본원인은
"하드코딩 24000 vs 파생 9600"의 2.5배 불일치였다(SoT 서두).

Dart `lib/core/pump_guard.dart` 포팅이되, **수치 정본 = 서버 TS**:
  - Dart 이관본의 validSyringeCapacitiesMl 4종 [1.25,0.5,2.5,5] → **9종**
    [0.025,0.05,0.1,0.25,0.5,1.0,1.25,2.5,5.0] (pumpGuard.ts VALID_SYRINGE_ML)으로 정정.
  - Tecan(Cavro XLP6000·XCalibur) 프리셋은 제거됨(2026-07-18 · 미도입·기기 미입고 — 서버 TS와 동일).
    현재 빌트인 = SY-01B 1종(+ custom). Dart 파일엔 남아 있으나 그건 동결된 포팅 오라클.
  - 그 외(프리셋 수치·custom 절대상한·기본값·단조성 2줄 순서)는 서버 TS와 대조 완료(동일).

라운딩(부록A P-8): round = half-up(양수 도메인 JS `Math.round` = `floor(x+0.5)`).
Python 내장 round() 는 banker's rounding 이라 **사용 금지** — `_round_half_up` 고정.
"""

from __future__ import annotations

import enum
import math
from dataclasses import dataclass
from typing import Any, Mapping


@dataclass(frozen=True, slots=True)
class PumpPreset:
    """PumpPreset 7필드 — SoT §6-1 (고정 필드명·타입·순서)."""

    pump_preset_id: str  # 항상 "sy01b" (사용자 지정·Tecan/Cavro 제거·2026-07-18)
    pump_full_stroke: int  # 풀스트로크
    pump_max_start_speed_hz: int  # v 상한(start speed)
    pump_max_top_speed_hz: int  # V 상한(top speed)
    pump_max_cutoff_speed_hz: int  # c 상한(cutoff speed)
    pump_max_slope: int  # L 상한(slope)
    pump_syringe_type_code: int  # 스톨 서브코드 U<code>


# 빌트인 프리셋 정식 수치표 — SoT §6-2 (입력 무시·강제 · 바이트 동일 SoT = pumpGuard.ts).
#
# 현재 빌트인 = SY-01B 1종. 사용자 지정(custom)·Tecan(Cavro)은 제거됨(2026-07-18 · 서버 pumpGuard.ts
# 와 동일) — clamp 는 어떤 입력이든 SY-01B 로 정규화한다. 도입 시 이 표에 1항목만 되살리면 된다.
PUMP_PRESETS: dict[str, PumpPreset] = {
    "sy01b": PumpPreset(
        pump_preset_id="sy01b",
        pump_full_stroke=12000,
        pump_max_start_speed_hz=1000,
        pump_max_top_speed_hz=6000,
        pump_max_cutoff_speed_hz=5400,
        pump_max_slope=20,
        pump_syringe_type_code=200,
    ),
}

# 유효 syringe 용량 이산값(mL) — v1.1.0 allowlist **9종**(서버 pumpGuard.ts VALID_SYRINGE_ML 정본).
VALID_SYRINGE_CAPACITIES_ML: frozenset[float] = frozenset(
    {0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 1.25, 2.5, 5.0}
)


def _round_half_up(x: float) -> int:
    """round = half-up(.5 올림 · 양수 도메인 JS Math.round 등가) — 부록A P-8. 내장 round() 금지."""
    return math.floor(x + 0.5)


def clamp_pump_preset(cfg: Mapping[str, Any] | None) -> PumpPreset:
    """clampPumpPreset(cfg) — SoT §6-3 (서버 ↔ pi 동일 알고리즘).

    **항상 SY-01B 로 정규화**한다(사용자 지정 제거·2026-07-18). 입력 pumpPresetId·수치
    (custom·unknown·레거시 포함)는 전부 무시하고 SY-01B 정식 수치를 강제 → 손 튜닝값이
    물리로 가는 경로 차단(과다흡입 안전). syringeCapacityMl 은 호출 측이 별도 주입한다.
    """
    return PUMP_PRESETS["sy01b"]


def resolve_syringe_capacity_ml(raw: Any, *, is_flavor: bool) -> float:
    """syringeCapacityMl 이산값 검증 — SoT §6-1 / O-15 (TS coerceSyringeCapacityMl 등가).

    유효집합(9종) 밖이면 **모드 기본값 폴백**(스냅 아님). 기본 용량 = 양 모드 공통 0.5mL
    (2026-07-17 확정: flavor 2펌프·fragrance 3펌프 모두 시린지 0.5mL). is_flavor 는
    시그니처 호환을 위해 유지하되 현재 폴백값은 동일하다.
    """
    fallback = 0.5
    if isinstance(raw, bool) or not isinstance(raw, (int, float)):
        return fallback
    d = float(raw)
    return d if d in VALID_SYRINGE_CAPACITIES_ML else fallback


@dataclass(frozen=True, slots=True)
class SyringeSpec:
    """부피→스텝 파생(SyringeSpec) — SoT §6-4 (하드코딩 금지·파생이 SoT).

      steps         = round( pumpFullStroke × volumeUl ÷ (syringeCapacityMl × 1000) )
      stepsPerMl    = pumpFullStroke ÷ syringeCapacityMl
      maxVolumeUl   = syringeCapacityMl × 1000   (per-pump 안전 게이트 상한)

    검산(§6-4): 12000 × 100 ÷ 500 = 2400 steps / 시린지 0.5mL stepsPerMl=24000, maxVol 500µL
               (양 모드 공통 0.5mL — 2026-07-17 확정: 식향 2펌프·향장향 3펌프). 참고: 1.25mL
               선택 시 12000 × 100 ÷ 1250 = 960 steps / stepsPerMl=9600.
    """

    pump_full_stroke: int
    syringe_capacity_ml: float

    @property
    def max_volume_ul(self) -> float:
        """per-pump 안전 게이트 상한(µL)."""
        return self.syringe_capacity_ml * 1000

    @property
    def steps_per_ml(self) -> float:
        """mL 당 스텝수."""
        return self.pump_full_stroke / self.syringe_capacity_ml

    def steps_for_volume_ul(self, volume_ul: float) -> int:
        """부피(µL) → 스텝수. round = half-up(양수 도메인·부록A P-8)."""
        return _round_half_up(
            self.pump_full_stroke * volume_ul / (self.syringe_capacity_ml * 1000)
        )

    # ── 초기화 파라미터 (용량 파생) — v1.1.0 `syringe_spec.dart` 포팅 ──────────────
    #
    # ⚠️ **모드가 아니라 용량이 결정한다.** v1.1.0 실기기 리포트에서 드러난 사고 경로가
    #    "설정 용량을 무시하고 모드 기본으로 초기화힘을 유도 → Z1R(Half)이어야 할 500µL
    #    시린지에 ZR(Full)이 나가는 매뉴얼 위반"이었다. 그래서 파생을 이 한 곳에 응집한다.
    #    v1.2.0 은 양 모드 공통 0.5mL 이므로 **둘 다 Z1R**(구 식향 1.25mL = ZR 이었다).

    @property
    def stall_current(self) -> int:
        """스톨 전류 단계 n (`U<code>,<n>R`) — 용량 파생 (Manual V1.2 §1.2 Table).

        ≤25µL → 4 · 50µL~1.25mL → 5 · 2.5~5mL → 6.
        """
        ul = self.max_volume_ul
        if ul <= 25:
            return 4
        if ul <= 1250:
            return 5
        return 6

    @property
    def init_command(self) -> str:
        """초기화 실행 명령 — 용량 파생 (Manual V1.2 §4.4.1). `Z<n1>R`: n1=0 Full·1 Half·2 Third.

        ≥1.0mL → `ZR`(Full) · 250·500µL → `Z1R`(Half) · 50·100µL → `Z2R`(Third).
        시린지 씰 보호가 목적이라 **작은 시린지에 Full 을 걸면 안 된다**.
        """
        if self.syringe_capacity_ml >= 1.0:
            return "ZR"
        if self.max_volume_ul >= 250:
            return "Z1R"
        return "Z2R"


def fragrance_ml_to_ul(amount_ml: float) -> float:
    """fragrance 단위 정규화 — SoT §6-6. amountMl → volumeUl(µL). 미스매치 = Code 11.

    (flavor volume 은 이미 µL 이므로 정규화 불필요.)
    """
    return amount_ml * 1000


def is_volume_within_gate(volume_ul: float, spec: SyringeSpec) -> bool:
    """recipe 스텝 검증 게이트 — SoT §6-4 / §9-1.
    0 < volumeUl ≤ maxVolumeUl. 위반 → CMD_VALIDATION_FAILED(drop).
    """
    return 0 < volume_ul <= spec.max_volume_ul


class EngineErrorClass(enum.Enum):
    """EnginePort 에러코드 분류 — SoT §6-7."""

    NORMAL = "normal"
    TRANSIENT = "transient"
    PERMANENT = "permanent"


def classify_engine_error_code(code: int) -> EngineErrorClass:
    """엔진 raw errorCode(정수) → 분류 — SoT §6-7.
    0 = 정상 / 1·7·11·15·timeout = transient(R=3 재시도) / 2·3·9·10 = permanent(즉시중단 FAILED).
    """
    if code == 0:
        return EngineErrorClass.NORMAL
    if code in (1, 7, 11, 15):
        return EngineErrorClass.TRANSIENT
    if code in (2, 3, 9, 10):
        return EngineErrorClass.PERMANENT
    # 미분류 코드는 보수적으로 permanent(안전측·즉시중단).
    return EngineErrorClass.PERMANENT


class StatusErrorCode(enum.Enum):
    """status.errorCode 7종 — SoT §6-7 / §9-2."""

    CMD_VALIDATION_FAILED = "CMD_VALIDATION_FAILED"
    DUPLICATE_DROPPED = "DUPLICATE_DROPPED"
    ENGINE_TIMEOUT = "ENGINE_TIMEOUT"
    ENGINE_ERROR_TRANSIENT = "ENGINE_ERROR_TRANSIENT"
    ENGINE_ERROR_PERMANENT = "ENGINE_ERROR_PERMANENT"
    PARTIAL_DISPENSE = "PARTIAL_DISPENSE"
    INTERRUPTED = "INTERRUPTED"

    @property
    def wire(self) -> str:
        return self.value

    @staticmethod
    def from_wire(v: Any) -> "StatusErrorCode | None":
        if not isinstance(v, str):
            return None
        for e in StatusErrorCode:
            if e.wire == v:
                return e
        return None
